### PR TITLE
Included MEPCurve ConnectionProperty Information

### LINKED ISSUE #911: Include connector information in MEPCurve convert
#### Description:
When the prototype convert for `MEPCurve` was created a few weeks ago the connector information was left out for simplicity purposes. Now is the time to tackle the issue and include it properly.

### PR DESCRIPTION
**Pull request by tiagogrossi**
### Issues addressed by this PR

Closes #911 

Adds the proper information to the `ConnectionProperty` to the `MEPCurve` converts. At the moment only `CableTray` is using the feature due to the lack of the property in `Duct` and `Pipe` objects.
Only the properties `IsStartConnected` and `IsEndConnected` have been assigned, the properties about the connection point have been ignored as they are irrelevant in this object and overall redundant.

### Test files
Link [here](https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/Revit_Toolkit/Revit_Toolkit-Issue911-IncludedMEPCurveConnectorInformation?csf=1&web=1&e=5Bjkit). Check is visual, in the GH script I've separated a few cases to check and the values expected.

### Changelog
Modified `BH.Revit.Engine.Core.Convert.CableTrayFromRevit`
Modified `BH.Revit.Engine.Core.Convert.DuctFromRevit`
Modified `BH.Revit.Engine.Core.Convert.PipeFromRevit`
Modified `BH.Revit.Engine.Core.Query.LocationCurve`

### CODE REVIEW

File: Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs:77]: Could we please avoid splitting a single statement into 3 lines? 

I would also consider rewriting the query to make the code a bit more readable:
```suggestion
            Vector revitCableTrayVector = ((revitCableTray.Location as LocationCurve)?.Curve as Line)?.Direction.VectorFromRevit().RoundCoordinates(4);
```

On top of that, this variable is not used until line 84, so could be declared later.

File: Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs:95]: I would consider migrating this whole part into a separate method, particularly in the case it could possibly be reused for other MEP types.

File: Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs:101]: This equality query is dangerous. I would rather use `BH.Engine.Geometry.Query.IsEqual` with tolerance.

File: Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs:165]: You can make use of the extension methods to make the code more readable:
```suggestion
            BH.oM.Geometry.Point referenceStart = curve.GetEndPoint(0).PointFromRevit();
```

File: Revit_Core_Engine/Convert/Environment/FromRevit/Duct.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Convert/Environment/FromRevit/Duct.cs:63]: Will `isStartConnected` and `isEndConnected` be relevant for this convert at any point? If not, I would consider adding an overload without `out` parameters that would call the one with them and simply not return them.

This comment refers to other converts as well.

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I went through the code as a first round of review - please see the code comments. Would be good to have some more discipline-aware reviewers involved too, to make sure that I am not derailing the concept 🙈 

@FraserGreenroyd @kayleighhoude

File: Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs
Comment (tiagogrossi) [reply Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs:95]: You're quite right, but I'd prefer not act on this if you're ok with it. Reason being that the other MEP obj don't take advantage of connector information. Also the only reason I'm doing this is to be more efficient in the path finding, since otherwise I'd have to query this info during it's runtime, so it saves time to grab it during the converts.

Comment (pawelbaran) [reply Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs:94]: Makes sense after the 2nd look at the code. Happy to keep it as is then, can be migrated out to a separate method in case of other converts wanting to use it too.

File: Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs:160]: Could you please embed this method inside `LocationCurveMEP`? I believe this should be a standard, not cabletray-specific.

File: Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Convert/Environment/FromRevit/CableTray.cs:166]: Is that a robust solution? What if the curves had some weird snail-like shape? Having said that, couldn't you joint the curves into a `PolyCurve`, flip it so that its start point is closer to the start point of element's location curve and then explode?

A bit of an overhead, but may be worth - please dismiss this comment if that does not make sense to you.

File: Revit_Core_Engine/Convert/Environment/FromRevit/Duct.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Convert/Environment/FromRevit/Duct.cs:60]: 👀

**Review by pawelbaran — CHANGES_REQUESTED** [review]
The code makes sense to me now, I would  only consider elevating `MatchRevitOrder` a bit - please see the code comments. Besides that good to go, let's wait for the functional review!

**Comment by tiagogrossi** [comment]
@pawelbaran thanks for the review. I've just commit the latest changes to address the comments. You'll notice that in the `CableTrayFromRevit` file, line 74, I've reversed to the manner I was retrieving the revit object vector from before (start-end.normalize), because for some reason `LocationCurve.Line.Direction` was retrieving some awkward results. Hope that's ok for you.

**Review by pawelbaran — COMMENTED** [review]
LGTM code wise 👍 However, I will keep this review to avoid getting the PR merged before the functionality gets approved by the discipline (@FraserGreenroyd @michaelhoehn @kayleighhoude).

**Review by michaelhoehn — APPROVED** [review]
Happy to approve after functionality tests with the provided scripts. Connection properties seem to correspond correctly 😄

**Review by pawelbaran — APPROVED** [review]

**Pull request merged**
